### PR TITLE
feat: dynamo-based ONNX export with subfunction support for CausalLM models

### PR DESCRIPTION
**Pull request by smedhe**
## Summary

This PR adds `use_dynamo=True` export support for CausalLM models, enabling `torch.export`-based ONNX export with optional repeated-subgraph subfunction extraction (`use_onnx_subfunctions=True`). It also includes a suite of robustness fixes to the ONNX transform pipeline discovered during integration.

  ### What changed

  **New infrastructure**
  - `QEfficient/utils/custom_op_utils.py`: `select_interface` helper that switches between the `torch.autograd.Function` path (TorchScript/eager) and `torch.ops.qefficient.*` path (dynamo) at call time, enabling the same model code to work under both exporters without code duplication.
  - `QEfficient/utils/export_utils.py`: `export_wrapper` refactored into four clearly separated export paths (plain TorchScript, TorchScript+subfunctions, flat dynamo, dynamo+subfunctions). `use_dynamo` is now an explicit parameter throughout rather than being re-read from `kwargs`. TorchScript monkey-patches are guarded behind `if not use_dynamo:` so they are never applied on the dynamo path. Added PyTorch version guard that raises a `RuntimeError` with install instructions when `use_dynamo=True` and torch < 2.13.
  - `QEfficient/utils/torch_patches.py`: Added `temporarily_disable_nested_compile_regions` context manager (Path 3) that unwraps static `@torch.compiler.nested_compile_region` decorators for flat-dynamo export, and `temporarily_enable_nested_compile_regions` reserved for future use.
  - `QEfficient/base/modeling_qeff.py`: 
    - `_prune_unused_fake_initializers` — removes `FakeTensor`-valued initializers that dynamo injects but are not referenced in the graph, preventing downstream ONNX load failures.
    - `_export` extended with `dynamic_shapes` parameter and full dynamo export path: sets `dynamo=True`, injects `custom_translation_table` mapping all `torch.ops.qefficient.*` ops to their ONNX counterparts, and reorders `example_inputs` / `dynamic_shapes` to match the model's `forward` signature before passing to `torch.onnx.export`.
  - `QEfficient/base/onnx_transforms.py`: Two new transforms for the dynamo+subfunctions path:
    - `PreserveNestedCacheRetainedStateTransform` — exposes CtxScatter outputs from repeated-subgraph function bodies as explicit ONNX outputs and renames KV cache graph inputs from `past_key.N_RetainedState` → `past_key.N` so the QAIC compiler can pair inputs and outputs correctly.
    - `RenameRepeatedSubgraphTransform` — renames `repeated_subgraphN` ONNX Local Functions to the actual decoder layer class names, with multi-pattern matching across known PyTorch naming conventions and `logger.warning` on mismatch or count mismatch.
  - `QEfficient/transformers/models/modeling_auto.py`: `convert_dynamic_axes_to_dynamic_shapes` converts ONNX `dynamic_axes` dicts to `torch.export` `Dim`-keyed `dynamic_shapes` dicts, handling nested `past_key_values`, compressed KV (DeepseekV3), and CB pool dimensions.
  - `QEfficient/utils/constants.py`: ONNX export opset bumped from **17 → 18**, required by the dynamo ONNX exporter.
  - `pyproject.toml`: Updated to torch **2.13.0** stable CPU wheels (minimum required for dynamo export); explicit per-python per-arch wheel URLs for x86_64 and aarch64. `onnxscript` updated to `0.6.2`, `compressed-tensors` updated to `0.17.0`.

  **CausalLM model wrappers** (`llama`, `mistral`, `qwen2`, `qwen3`, `gemma`, `gemma2`, `gpt2`, `gpt_bigcode`, `gptj`, `falcon`, `mpt`, `phi`, `phi3`, `codegen`, `starcoder2`, `granite`, `granitemoe`, `mixtral_moe`, `qwen3_moe`, `olmo2`, `glm4_moe`, `grok_1`, `deepseek_v3`, `gpt_oss`, `llama_swiftkv`):
  - `QEffDecoderLayer.forward` decorated with `@torch.compiler.nested_compile_region` so dynamo emits repeated-subgraph ONNX Local Functions instead of inlining each layer individually.
  - `torch.tensor(MIN_MASKED_ATTENTION_VALUE)` replaced with `torch.full_like(attn_weights, ...)` throughout — removes a scalar constant that dynamo cannot trace as a dynamic value and that caused shape inference failures during export.
  - `select_interface` wired into all custom op call sites (`CtxScatter`, `CtxGather`, `CustomRMSNorm`, and CB/3D variants) so both the TochTrace and dynamo paths dispatch to the correct backend op without branching in model code.

### CODE REVIEW

**Comment by quic-amitraj** [comment]
Please clean the dead and commented code.

File: QEfficient/base/modeling_qeff.py [hunk 1]
```diff
@@ -55,6 +73,19 @@
 from QEfficient.utils.export_utils import export_wrapper
 from QEfficient.utils.torch_patches import layerwise_safe_onnx_export_patches
 
+
+def _prune_unused_fake_initializers(onnx_program) -> None:
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:77]: nit: Can we move these methods to onnx utils? we should not be keeping these methods in base classes

File: QEfficient/base/modeling_qeff.py [hunk 2]
```diff
@@ -433,19 +466,98 @@ def _resolve_pkv_names(layer_idx, layer_state):
             dynamic_axes = {rename_map.get(k, k): v for k, v in dynamic_axes.items()}
             input_names = aligned_input_names
 
+        # Rearrange example_inputs and dynamic_shapes to follow model.forward signature
+        sig = inspect.signature(self.model.forward)
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:470]: better to keep a seperate method for dynamo based export rather than adding it part of the current method. This will make things scalable.

File: QEfficient/base/modeling_qeff.py [hunk 3]
```diff
@@ -433,19 +466,98 @@ def _resolve_pkv_names(layer_idx, layer_state):
...
+
+        ordered_example_inputs = OrderedDict()
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:472]: from python 3.8 the built in dict preserves insertion orde, Any specific reason why OrderedDict is being used here?

File: QEfficient/base/modeling_qeff.py [hunk 4]
```diff
@@ -1080,6 +1198,7 @@ def _compile(
                 command.append(f"-custom-IO-list-file={custom_io_yaml}")
 
         command.append(f"-aic-binary-dir={qpc_path}")
+        print(command)
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:1201]: move it to logger

File: QEfficient/base/modeling_qeff.py [hunk 5]
```diff
@@ -433,19 +466,98 @@ def _resolve_pkv_names(layer_idx, layer_state):
...
+            export_kwargs = {} if export_kwargs is None else export_kwargs
+
+            if use_dynamo:
+                dynamic_axes = None
+                export_kwargs = dict(export_kwargs)
+                export_kwargs.setdefault("report", False)
+                export_kwargs.setdefault("optimize", False)
+                export_kwargs["dynamo"] = True
+                export_kwargs["custom_translation_table"] = {
+                    **(export_kwargs.pop("custom_translation_table", None) or {}),
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:506]: Can we create a common custom op map somewhere and use it everywhere? this would create a dependency of update the custom op here as well when we add a new custom op. Im open for discussion on this.

**Comment by vbaddi** [comment]
duplicate #1178
Pls continue the implementation in #1178 thanks.
Closing this.

File: QEfficient/base/onnx_transforms.py [hunk 6]
```diff
@@ -255,6 +271,236 @@ def apply(cls, model: ModelProto, layer_idx=0) -> bool:
...
+            # etc.), so we sort by the scatter output name instead — dynamo
+            # preserves "key"/"value" in output tensor names even when input
+            # argument names are opaque.
+            scatter_nodes = [
+                fn_node for fn_node in fn.node if fn_node.op_type in cls._SCATTER_OP_TYPES and fn_node.output
+            ]
+            if len(scatter_nodes) != 2:
+                continue
+
+            def _scatter_sort_key(n):
```
Comment (quic-rishinr) [inline QEfficient/base/onnx_transforms.py:329]: move the function declaration outside of the loop.

File: QEfficient/base/onnx_transforms.py [hunk 7]
```diff
@@ -255,6 +271,236 @@ def apply(cls, model: ModelProto, layer_idx=0) -> bool:
...
+            if fn is None:
+                continue
+
+            # Collect scatter nodes that write back the KV cache.
+            # Sort by first-input name so the key scatter reliably precedes the
+            # value scatter: dynamo names function-body args generically (arg7_1
+            # etc.), so we sort by the scatter output name instead — dynamo
+            # preserves "key"/"value" in output tensor names even when input
+            # argument names are opaque.
+            scatter_nodes = [
```
Comment (quic-rishinr) [inline QEfficient/base/onnx_transforms.py:323]: there are 4 loops for fetching and updating hte scatter nodes can we keep one loop for it and keep it simple?

File: QEfficient/customop/ctx_scatter_gather.py [hunk 8]
```diff
@@ -217,7 +217,7 @@ def symbolic(g: torch.Graph, data: torch.Value, ctx_indices: torch.Value) -> tor
 
 @onnxscript.script(onnxscript.values.Opset("com.qualcomm.cloud", 1))
 def CtxGather(
-    data: onnxscript.FLOAT, ctx_indices: onnxscript.INT32, comp_ctx_len: onnxscript.INT32
+    data: onnxscript.FLOAT, ctx_indices: onnxscript.INT32, comp_ctx_len: onnxscript.INT64
```
Comment (quic-rishinr) [inline QEfficient/customop/ctx_scatter_gather.py:220]: why we are updating it to int64?

File: QEfficient/transformers/models/glm4_moe/modeling_glm4_moe.py [hunk 9]
```diff
@@ -349,7 +362,7 @@ def forward(
 
         if sin_cached is not None and cos_cached is not None:
             sin, cos = sin_cached, cos_cached
-            rotary_dim = int(self.rotary_emb.cos_cached.shape[-1])
+            rotary_dim = int(self.rotary_emb.cos_cached.detach().clone().shape[-1])
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/glm4_moe/modeling_glm4_moe.py:365]: do we need to have a clone for rotary_emb?

File: QEfficient/transformers/models/modeling_auto.py [hunk 10]
```diff
@@ -1559,6 +1559,7 @@ def export(
                 kv_offload=True,
                 continuous_batching=self.continuous_batching,
                 comp_ctx_lengths=self.comp_ctx_lengths_decode,
+                prefill_seq_len=prefill_seq_len,
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py:1562]: is this a rebase issue? this was removed from mainline couple of days back

File: QEfficient/transformers/models/modeling_auto.py [hunk 11]
```diff
@@ -3720,6 +3721,122 @@ def _factory(model_id, config):
...
+        """
+        Convert ONNX dynamic_axes format to torch.export dynamic_shapes format
+
+        Args:
+            dynamic_axes: ONNX format like {"input_ids": {0: "batch_size", 1: "seq_len"}}
+
+        Returns:
+            dynamic_shapes: torch.export format with Dim objects matching model forward args
+        """
+        from torch.export import Dim
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py:3734]: move the imports above

File: QEfficient/transformers/models/modeling_auto.py [hunk 12]
```diff
@@ -3720,6 +3721,122 @@ def _factory(model_id, config):
             final_compile=final_compile,
         )
 
+    def convert_dynamic_axes_to_dynamic_shapes(self, dynamic_axes: Dict[str, Dict[int, str]]) -> Dict[str, any]:
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py:3724]: refactor the method. 
1. Move the nested helper method out of convert_dynamic_axes_to_dynamic_shapes.
2. Replace the magic numbers with well named constants (or move them to a common utils/constants module).
3. Consider moving this logic to the export utilities instead of keeping it inside the AutoModel class
4. Remove the redundant iterations over dynamic_axes.items() and consolidate the logic where possible. for now there are 3 loops over  dynamic_axes.items()

File: QEfficient/transformers/models/glm4_moe/modeling_glm4_moe.py (hunk 9 above)
Comment (smedhe) [reply QEfficient/transformers/models/glm4_moe/modeling_glm4_moe.py:365]: needed for dynamo aliasing issue while graph tracing

File: QEfficient/customop/ctx_scatter_gather.py (hunk 8 above)
Comment (smedhe) [reply QEfficient/customop/ctx_scatter_gather.py:220]: for some models this was giving a dtype issue in compile
